torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -0,0 +1,201 @@
+#include <torch/csrc/jit/codegen/cuda/lower_predicate.h>
+
+#include <torch/csrc/jit/codegen/cuda/index_compute.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+#include <torch/csrc/jit/codegen/cuda/predicate_compute.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+class TORCH_CUDA_CU_API ConditionalFromPredicateModifier {
+ public:
+  ConditionalFromPredicateModifier(Fusion* fusion) {
+    p2c_root_map_ = loop_utils::p2cRootMap(fusion->exprs());
+  }
+
+  void process(const std::vector<kir::Expr*>& exprs) {
+    FUSER_PERF_SCOPE("ConditionalFromPredicateModifier::process");
+    for (auto* expr : exprs) {
+      handle(expr);
+    }
+  }
+
+  const std::unordered_map<kir::Expr*, kir::Expr*>& replacementMap() const {
+    return expr_replacement_map_;
+  }
+
+ private:
+  void handle(kir::Expr* expr) {
+    if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      handle(for_loop);
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      handle(ite);
+    } else if (expr != nullptr && expr->predicate() != nullptr) {
+      // Replace expr predicate with bool conditional
+      auto conditional = generateConditional(expr->predicate());
+      expr->predicate()->setValue(conditional);
+      TORCH_INTERNAL_ASSERT(expr->predicate()->value() != nullptr);
+    }
+  }
+
+  void handle(kir::ForLoop* fl) {
+    for_loops_structure_.push_back(fl);
+
+    const auto exprs_copy = fl->body().exprs();
+    for (auto expr : exprs_copy) {
+      handle(expr);
+    }
+
+    for_loops_structure_.pop_back();
+  }
+
+  void handle(kir::IfThenElse* ite) {
+    TORCH_INTERNAL_ASSERT(ite->cond() != nullptr);
+
+    // If ite already has Bool conditional, handle internal expressions
+    // Otherwise, generate conditional and update predicate
+    if (ite->cond()->hasValue()) {
+      const auto then_exprs_copy = ite->thenBody().exprs();
+      for (auto expr : then_exprs_copy) {
+        handle(expr);
+      }
+
+      const auto else_exprs_copy = ite->elseBody().exprs();
+      for (auto expr : else_exprs_copy) {
+        handle(expr);
+      }
+    } else {
+      auto conditional = generateConditional(ite->cond());
+      TORCH_INTERNAL_ASSERT(conditional != nullptr);
+      TORCH_INTERNAL_ASSERT(conditional->isA<kir::Bool>());
+
+      // If predicate is not required
+      if (conditional->isConst() && conditional->value().value()) {
+        TORCH_INTERNAL_ASSERT(ite->elseBody().empty());
```

Comment:
Why assert here? If it's just true could you simply remove the ITE, erase the else, or do nothing?

Reply:
Ah okay, seems like to remove the ITE you're assuming it's an inline predicate that ends up being `true`. That way you can just replace the ITE with the only Expr in it. I think this could be further generalized but seems unnecessary to do so in this PR.

Reply:
The assertion doesn't seem to work with shift and padding. A shift predicate has a non-empty else block, which contains the padding block. If `conditional` is true, it's safe to ignore the padding block. Do we really need this assertion?

Reply:
Is the shift predicate ever the `true` value? IIUC, it always some boolean expression.

This IfThenElse statement is removed when the predicate is true:
```
if (true) {
  // expression
} else {
  // never visited
}
```

Reply:
In a follow-up PR, I extended the first shift PR with additional support of broadcast and reduction (#870). 

In `getInlinePredicate`, when the expr just initializes a reduction buffer on local memory, we just return `trueVal()`. In the follow-up PR, I do the same thing when a tensor has halo, so in that case the `shift` predicate is just `true`. 

`FusionShiftReduction1` hits this. Here's the generated code (with the assertion deleted):

```
__global__ void kernel1(Tensor<float, 2> T0, Tensor<float, 1> T3) {
  #pragma unroll 1
  for(size_t ki45 = 0; ki45 < (ceilDiv(T0.size[0], 4)); ++ki45) {
    float T2[(4 + 1)];
    #pragma unroll 1
    for(size_t ki60 = 0; ki60 < (4 + 1); ++ki60) {
      T2[ki60] = 0;
    }
    #pragma unroll 1
    for(size_t ki51 = 0; ki51 < (4 + 1); ++ki51) {
      #pragma unroll 1
      for(size_t ki53 = 0; ki53 < T0.size[1]; ++ki53) {
        float T1[1];
        if ((((((ki45 * 4) + ki51) >= 1) && (((ki45 * 4) + ki51) < (T0.size[0] + 1))) && (ki53 < T0.size[1]))) {
          T1[0]
            = T0[((((ki45 * 4) + ki51) + -1) * T0.stride[0]) + (ki53 * T0.stride[1])]
            + (float) 1;
        } else {
          if (((((ki45 * 4) + ki51) < (T0.size[0] + 1)) && (ki53 < T0.size[1]))) {
            T1[0] = 0;
          }
        }
        if ((((((ki45 * 4) + ki51) >= 1) && (((ki45 * 4) + ki51) < (T0.size[0] + 1))) && (ki53 < T0.size[1]))) {
          T2[ki51]
            = T2[ki51]
            + T1[0];
        } else {
          if (((((ki45 * 4) + ki51) < (T0.size[0] + 1)) && (ki53 < T0.size[1]))) {
            T2[ki51] = 0;
          }
        }
      }
    }
    #pragma unroll 1
    for(size_t ki47 = 0; ki47 < 4; ++ki47) {
      if (((((ki45 * 4) + ki47) >= 1) && (((ki45 * 4) + ki47) < T0.size[0]))) {
        T3[(((ki45 * 4) + ki47) * 1)]
           = T2[ki47];
      } else {
        if ((((ki45 * 4) + ki47) < T0.size[0])) {
          T3[(((ki45 * 4) + ki47) * 1)] = 0;
        }
      }
    }
  }
}
```

The initialization of `T2` is the one that would hit the assertion.

Reply:
I can remove the assertion. I was confused why there would be an Else block if the predicate is `true`.

Reply:
Thanks. That's because the if-then-else structure is created before the predicate is filled by `true`.

Reply:
I pushed the update.